rarible-subgraph/src/utils.ts

```diff
@@ -0,0 +1,154 @@
+  export const LENDING = "LENDING";
+  export const YIELD = "YIELD";
+  export const BRIDGE = "BRIDGE";
+  export const DAO = "DAO";
+  export const NFT_MARKET_PLACE = "NFT_MARKET_PLACE";
+  export const STAKING = "STAKING";
+  export const P2E = "P2E";
+  export const LAUNCHPAD = "LAUNCHPAD";
+}
+
+export namespace AirProtocolActionType {
+  export const ALL = "ALL";
+  export const BUY = "BUY";
+  export const SELL = "SELL";
+  export const MINT = "MINT";
+  export const BURN = "BURN";
+  export const ATTEND = "ATTEND";
+  export const EARN = "EARN";
+  export const SWAP = "SWAP";
+  export const ADD_LIQUIDITY = "ADD_LIQUIDITY";
+  export const REMOVE_LIQUIDITY = "REMOVE_LIQUIDITY";
+  export const ADD_TO_FARM = "ADD_TO_FARM";
+  export const REMOVE_FROM_FARM = "REMOVE_FROM_FARM";
+  export const CLAIM_FARM_REWARD = "CLAIM_FARM_REWARD";
+  export const LEND = "LEND";
+  export const BORROW = "BORROW";
+  export const FLASH_LOAN = "FLASH_LOAN";
+  export const STAKE = "STAKE";
+  export const RESTAKE = "RESTAKE";
+  export const UNSTAKE = "UNSTAKE";
+  export const DELEGATE = "DELEGATE";
+  export const CLAIM_REWARDS = "CLAIM_REWARDS";
+}
+
+class RoyaltyDetails {
+  royaltyAmounts: BigInt[];
+  royaltyRecipients: Address[];
+};
+
+export function getRoyaltyDetails(
+  tokenId: BigInt,
+  tokenAddress: Address,
+  restValue: BigInt,
+  amount: BigInt,
+): RoyaltyDetails {
+  // extract data from contract logic comes here
+  let contractInstance = HasSecondarySaleFees.bind(tokenAddress);
+  let supportsInterface = contractInstance.try_supportsInterface(INTERFACE_ID_FEES);
+
+  if (!supportsInterface.reverted && supportsInterface.value) {
+    let royaltyRecipients = contractInstance.getFeeRecipients(tokenId);
+    let royaltyAmounts = contractInstance.getFeeBps(tokenId);
+    for (var i = 0; i < royaltyAmounts.length; i++) {
+      let subFeeResponse = subFeeInBp(restValue, amount, royaltyAmounts[i]);
+      royaltyAmounts[i] = subFeeResponse.realFee;
+      restValue = subFeeResponse.newValue;
+    }
+    return {
+      royaltyAmounts,
+      royaltyRecipients,
+    };
+  };
+
+  return {
+    royaltyAmounts: [],
+    royaltyRecipients: [],
+  };
+}
+
+class SubFeeResponse {
+  newValue: BigInt;
+  realFee: BigInt;
+}
+
+function subFeeInBp(
+  value: BigInt,
+  total: BigInt,
+  feeInBp: BigInt,
+): SubFeeResponse {
+  return subFee(value, bp(total, feeInBp));
+}
+
+function subFee(
+  value: BigInt,
+  fee: BigInt
+): SubFeeResponse {
+  let newValue: BigInt;
+  let realFee: BigInt;
+  if (value > fee) {
+    newValue = value.minus(fee);
+    realFee = fee;
+  } else {
+    newValue = BigInt.fromI32(0);
+    realFee = value;
+  }
+  return { newValue, realFee };
+}
+
+function bp(value1: BigInt, value2: BigInt): BigInt {
+  return value1.times(value2).div(BigInt.fromI32(10000));
+}
+
+class BeneficiaryDetails {
+  beneficiaryFee: BigInt;
+  beneficiary: Address;
+  restValue: BigInt;
+}
+
+export function getFeeBeneficiaryDetails(
```

Comment:
I would recommend to write math in function document so that if we revisit the code after 6 months, it possible to understand it. Also do same for other functions.

Reply:
resolved in https://github.com/Airstack-xyz/Subgraphs/pull/20